### ISSUE TITLE
Make legacy resource ordering customizable in the kustomization.yaml file

### ISSUE THREAD

**Issue by yanniszark** [kubernetes-sigs/kustomize#3913]
**Is your feature request related to a problem? Please describe.**

Related issue: https://github.com/kubernetes-sigs/kustomize/issues/3794

TLDR:
- Kustomize changed the order it prints its resources.
- The new order broke Kubeflow and we can't upgrade to v4+ (https://github.com/kubeflow/manifests/issues/1797#issuecomment-817017412)
- Order can't be reversed because it's been in effect for 1 year now, so changing it will break users.
- Instead, we can make the ordering configurable via the `kustomization.yaml` file, as described in https://github.com/kubernetes-sigs/kustomize/issues/3794#issuecomment-819942448

**Describe the solution you'd like**

Make resource ordering configurable via the `kustomization.yaml` file.
The sorter accepts the `orderFirst` and `orderLast` arguments, to know which resources to order first and last.
Then we need to decide what to call the new field for the sorter in the `kustomization.yaml` file.

cc @monopole

**Comment by KnVerey** [comment]
I think that this new field should **replace** the `--reorder` flag. Because it changes the output, this flag is currently violating Kustomize's principle of having [no build-time side-effects](https://kubectl.docs.kubernetes.io/faq/kustomize/eschewedfeatures/#build-time-side-effects-from-cli-args-or-env-variables). IMO we should add a deprecation warning to the flag when this new field merges, and then remove the flag from a future major version release. While both are present, the field should take precedence (if you've switched your kustomizations to the new way, we should ignore the old)... though this might be complicated in practice because the flag's scope is global and the field's should not be.

Maybe something like this?

```yaml
sortOrder: {legacy,fifo} # expandable e.g. if we want to add a non-legacy option that provides semantically stable ordering but without the release order implications
legacySortOptions: #  it's an error if this is present and sortOrder isn't legacy
  orderFirst:
  - {GVK}
  orderLast:
  - {GVK}
```

Or just the latter field, and stock legacy ordering would be enabled with `legacySortOptions: {}`. I don't love the clutter of two new top-level fields, but the latter reads awkwardly to me.

@yanniszark wdyt of the above? Are you interested in contributing this field?

**Comment by yanniszark** [comment]
Hi @KnVerey,
I think the discussion about kustomize removing the reorder flag is very interesting and something we should consider after this feature lands. We (Arrikto, Kubeflow) are interested in pushing this feature forward, so I will start with the example you suggested as my goal.

**Comment by numbsafari** [comment]
Brought here by some problems deploying cert-manager together with some other tools using kustomize. I think this idea of having a logical default ordering, but the ability on a per-kustomization basis to describe a desired ordering based on resource type makes a lot of sense. The main reason I like it is that you can memorialize/document the need for the ordering within your data/files, rather than relying on an external side-effect registered via a CLI flag. In addition, given the existence of CustomResourceDefinitions with their own potential semantics, it's impossible for kustomize to know "everything" it needs to know.

**Comment by natasha41575** [comment]
>  I don't love the clutter of two new top-level fields, but the latter reads awkwardly to me.

Would it make sense to consolidate it into a single top-level field with nested subfields?

```
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization

resources:
- ../base

sortOptions:
  sortOrder: legacy
  legacySortOptions:
    orderFirst:
    - {GVK}
    orderLast:
    - {GVK}
```

@KnVerey @yanniszark WDYT?

**Comment by yanniszark** [comment]
@natasha41575 IIUC, you are proposing this:
```
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization

resources:
- ../base

sortOptions:
  legacy:
    orderFirst:
    - {GVK}
    orderLast:
    - {GVK}
```


VS

```yaml
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization

resources:
- ../base

sortOptions:
  sortOrder: legacy
  legacySortOptions:
    orderFirst:
    - {GVK}
    orderLast:
    - {GVK}
```

TBF, the second one seems more intuitive to me. @KnVerey what do you think?

**Comment by natasha41575** [comment]
@KnVerey please let us know your thoughts

**Comment by yanniszark** [comment]
ping @KnVerey 
@natasha41575 I will start the implementation of the latter in the PR, while we wait.
Hopefully, there will only be small changes on top of that :)

**Comment by KnVerey** [comment]
That LGTM. I'd suggest dropping `sortOrder` to just `order` since it is under `sortOptions`, but I don't feel strongly about it.

**Comment by natasha41575** [comment]
>  I'd suggest dropping sortOrder to just order since it is under sortOptions, but I don't feel strongly about it.

+ 1

**Comment by KnVerey** [comment]
https://github.com/kubernetes-sigs/kustomize/pull/4019 is back in active development and getting really close now

/remove-lifecycle stale
